pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,625 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "MultiTopicsConsumerImpl.h"
+
+DECLARE_LOG_OBJECT()
+
+namespace pulsar {
+MultiTopicsConsumerImpl::MultiTopicsConsumerImpl(ClientImplPtr client, const std::vector<std::string>& topics,
+                                                 const std::string& subscriptionName, TopicNamePtr topicName,
+                                                 const ConsumerConfiguration& conf,
+                                                 const LookupServicePtr lookupServicePtr)
+    : client_(client),
+      subscriptionName_(subscriptionName),
+      topic_(topicName ? topicName->toString() : "EmptyTopics"),
+      conf_(conf),
+      state_(Pending),
+      messages_(1000),
+      listenerExecutor_(client->getListenerExecutorProvider()->get()),
+      messageListener_(conf.getMessageListener()),
+      namespaceName_(topicName ? topicName->getNamespaceName() : boost::shared_ptr<NamespaceName>()),
+      lookupServicePtr_(lookupServicePtr),
+      allTopicPartitionsNumber_(boost::make_shared<std::atomic<int>>(0)),
+      topics_(topics) {
+    std::stringstream consumerStrStream;
+    consumerStrStream << "[Muti Topics Consumer: "
+                      << "TopicName - " << topic_ << " - Subscription - " << subscriptionName << "]";
+    consumerStr_ = consumerStrStream.str();
+
+    if (conf.getUnAckedMessagesTimeoutMs() != 0) {
+        unAckedMessageTrackerPtr_.reset(
+            new UnAckedMessageTrackerEnabled(conf.getUnAckedMessagesTimeoutMs(), client, *this));
+    } else {
+        unAckedMessageTrackerPtr_.reset(new UnAckedMessageTrackerDisabled());
+    }
+}
+
+void MultiTopicsConsumerImpl::start() {
+    if (topics_.empty()) {
+        setState(Ready);
+        LOG_INFO("No topics passed in when create MultiTopicsConsumer.");
+        multiTopicsConsumerCreatedPromise_.setValue(shared_from_this());
+        return;
+    }
+
+    // start call subscribeOneTopicAsync for each single topic
+    boost::shared_ptr<std::atomic<int>> atomicIntPtr = boost::make_shared<std::atomic<int>>(0);
+
+    int topicsNumber = topics_.size();
+    // subscribe for each passed in topic
+    for (std::vector<std::string>::const_iterator itr = topics_.begin(); itr != topics_.end(); itr++) {
+        subscribeOneTopicAsync(*itr).addListener(
+            boost::bind(&MultiTopicsConsumerImpl::handleOneTopicSubscribe, shared_from_this(), _1, _2, *itr,
+                        atomicIntPtr, topicsNumber));
+    }
+}
+
+void MultiTopicsConsumerImpl::handleOneTopicSubscribe(Result result, Consumer consumer,
```

Comment:
This is a callback, but the name doesn't indicate it.
handleOneTopicSubscribe -> handleOneTopicSubscribed